File: my-code/290/main.go
package main

import (
	"strings"
)

// @Time     : 2021/5/30 22:27
// @Author   : Ranshi
// @File     : 290. 单词规律.go
func wordPattern(pattern string, s string) bool {
	mapping1 := make(map[uint8]string)
	mapping2 := make(map[string]uint8)
	strList := strings.Split(s, " ")
	if len(strList) != len(pattern) {
		return false
	}
	for i, str := range strList {
		if _, ok := mapping1[pattern[i]]; ok {
			if str != mapping1[pattern[i]] {
				return false
			}
		} else {
			mapping1[pattern[i]] = str
		}
		if _, ok := mapping2[str]; ok {
			if pattern[i] != mapping2[str] {
				return false
			}
		} else {
			mapping2[str] = pattern[i]
		}
	}
	return true
}

File: my-code/290/main.py
# coding: utf-8
# @Time     : 2021/08/11 11:12
# @Author   : Ranshi
# @File     : 123.py


class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        mapping, rev_mapping = {}, {}
        chs_l = s.split(" ")
        if len(chs_l) != len(pattern):
            return False
        for i in range(len(pattern)):
            if chs_l[i] in mapping and pattern[i] in rev_mapping:
                if mapping[chs_l[i]] != pattern[i] or rev_mapping[pattern[i]] != chs_l[i]:
                    return False
            elif chs_l[i] not in mapping and pattern[i] not in rev_mapping:
                mapping[chs_l[i]] = pattern[i]
                rev_mapping[pattern[i]] = chs_l[i]
            else:
                return False
        return True


if __name__ == "__main__":
    print(Solution().wordPattern(pattern="abba", s="dog cat cat dog"))
